File: app/src/main/java/com/example/pomoapp/pomoActivity.java
package com.example.pomoapp;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.net.UrlQuerySanitizer;
import android.os.Build;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.os.PersistableBundle;
import android.os.StrictMode;
import android.renderscript.ScriptGroup;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;


import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NotificationBuilderWithBuilderAccessor;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Locale;

public class pomoActivity extends AppCompatActivity {
    private final String CHANNEL_ID = "Channel1";
    private final int NOTIFICATION_ID = 001;
    // 25 minutes to milliseconds
    private static final long START_TIME_IN_MILL = 12000;

    private TextView textViewCountdown;
    private Button buttonStartPause;
    private Button buttonReset;
    private CountDownTimer countDownTImer;
    private boolean timerRunning;
    private long timeLeftInMill = START_TIME_IN_MILL;
    private NotificationManagerCompat notificationManager;

    private static HttpURLConnection connection;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.pomo_activity);
        buttonStartPause = findViewById(R.id.startPomo);
        buttonReset = findViewById(R.id.resetPomo);
        textViewCountdown = findViewById(R.id.text_view_countdown);


        buttonStartPause.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (timerRunning) {
                    pauseTimer();
                } else {
                    startTimer();
                }
            }
        });
        buttonReset.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                resetTimer();
            }
        });
        updateCountDownText();
        notificationManager = NotificationManagerCompat.from(this);
        createNotificationChannels();

    }

    private static String requestQuote() throws Exception {
        URL url = new URL("https://quotes.rest/qod?category=inspire");
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("GET");
        InputStream input = connection.getInputStream();
        String quote = input.toString();
        return quote;
    }
    private void startTimer() {
       countDownTImer = new CountDownTimer(timeLeftInMill, 1000) {
           @Override
           public void onTick(long millisUntilFinished) {
               timeLeftInMill = millisUntilFinished;
               updateCountDownText();
           }

           @Override
           public void onFinish() {
                timerRunning = false;
                buttonReset.setText("Reset");
                buttonStartPause.setText("Start");
                // set up notification here.
               sendNotificationChannel();

           }
       }.start();
        timerRunning = true;
        buttonStartPause.setText("Pause");
    }
    private void pauseTimer() {
        countDownTImer.cancel();
        timerRunning = false;
        buttonStartPause.setText("Resume");

    }
    private void resetTimer() {
        timeLeftInMill = START_TIME_IN_MILL;
        updateCountDownText();

    }
    private void updateCountDownText() {
        int minute = (int) (timeLeftInMill / 1000) / 60;
        int seconds = (int) (timeLeftInMill / 1000) % 60;

        String timeLeftFormatted = String.format(Locale.getDefault(),"%02d:%02d", minute, seconds);
        textViewCountdown.setText(timeLeftFormatted);

    }
    private void sendNotificationChannel() {
        Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_timer_notification)
                .setContentTitle("Timer is complete")
                .setContentText("take a quick break, you deserve it!")
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setCategory(NotificationCompat.CATEGORY_MESSAGE)
                .build();

        notificationManager.notify(1, notification);
    }
    private void createNotificationChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel1 = new NotificationChannel(
                    CHANNEL_ID,
                    "Channel 1",
                    NotificationManager.IMPORTANCE_DEFAULT
            );
            channel1.setDescription("this is channel 1");
            NotificationManager manager = getSystemService(NotificationManager.class);
            manager.createNotificationChannel(channel1);

        }

    }
}

File: app/src/main/java/com/example/pomoapp/pomoActivity.kt
package com.example.pomoapp

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.CountDownTimer
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import java.util.*

class pomoActivity : AppCompatActivity() {
    private val CHANNEL_ID = "Channel1"
    private val NOTIFICATION_ID = 1
    private var textViewCountdown: TextView? = null
    private var buttonStartPause: Button? = null
    private var buttonReset: Button? = null
    private var countDownTImer: CountDownTimer? = null
    private var timerRunning = false
    private var timeLeftInMill = START_TIME_IN_MILL
    private var notificationManager: NotificationManagerCompat? = null
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.pomo_activity)
        buttonStartPause = findViewById(R.id.startPomo)
        buttonReset = findViewById(R.id.resetPomo)
        textViewCountdown = findViewById(R.id.text_view_countdown)
        val buttonStartPause = buttonStartPause
        val buttonReset = buttonReset
        buttonStartPause?.setOnClickListener(View.OnClickListener {
            if (timerRunning) {
                pauseTimer()
            } else {
                startTimer()
            }
        })

        buttonReset?.setOnClickListener(View.OnClickListener { resetTimer() })
        updateCountDownText()
        notificationManager = NotificationManagerCompat.from(this)
        createNotificationChannels()
    }

    private fun startTimer() {
        countDownTImer = object : CountDownTimer(timeLeftInMill, 1000) {
            override fun onTick(millisUntilFinished: Long) {
                timeLeftInMill = millisUntilFinished
                updateCountDownText()
            }

            override fun onFinish() {
                timerRunning = false
                buttonReset!!.text = "Reset"
                buttonStartPause!!.text = "Start"
                // set up notification here.
                sendNotificationChannel()
            }
        }.start()
        timerRunning = true
        buttonStartPause!!.text = "Pause"
    }

    private fun pauseTimer() {
        countDownTImer!!.cancel()
        timerRunning = false
        buttonStartPause!!.text = "Resume"
    }

    private fun resetTimer() {
        timeLeftInMill = START_TIME_IN_MILL
        updateCountDownText()
    }

    private fun updateCountDownText() {
        val minute = (timeLeftInMill / 1000).toInt() / 60
        val seconds = (timeLeftInMill / 1000).toInt() % 60
        val timeLeftFormatted = String.format(Locale.getDefault(), "%02d:%02d", minute, seconds)
        textViewCountdown!!.text = timeLeftFormatted
    }

    private fun sendNotificationChannel() {
        // use val instead of var since it will be immutable then
        val notification = NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_timer_icon)
                .setContentTitle("Timer is Complete")
                .setContentText("take a 5-minute break, you deserve it:) resume the timer after you're done!")
                .setStyle(NotificationCompat.BigTextStyle()
                        .bigText("take a 5-minute break, you deserve it:) resume the timer after you're done!"))
                // set priority to HIGH so that notification will be displayed at top of screen
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setCategory(NotificationCompat.CATEGORY_MESSAGE)
                // set the intent that will fire when the user taps the notification
                .setContentIntent(PendingIntent.getActivity(this, 0, intent, 0))
                .setAutoCancel(true)    // notification is removed when tapped by user
                .setVibrate(longArrayOf(1L, 2L, 3L))    // add vibration to notification
                .build()
        notificationManager!!.notify(1, notification)
    }

    // Create notification channel, including setting priority of notification
    private fun createNotificationChannels() {
        // if statement that checks to make sure OS is API 26 or greater
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notifChannel1 = NotificationChannel(
                    CHANNEL_ID,
                    "Channel 1",
                    NotificationManager.IMPORTANCE_DEFAULT
            )
            notifChannel1.description = "This is Channel 1"
            val notificationManager:NotificationManager =
                    getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.createNotificationChannel(notifChannel1)
        }
    }

    companion object {
        // 25 minutes to milliseconds
        private const val START_TIME_IN_MILL: Long = 18000
    }
}